File: Jawa_cavé/Cafe.java
// -*- coding: utf-8 -*-
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;

public class Cafe {

    // ── Constantes ────────────────────────────────────────────────────────────
    static final String ARQUIVO_CSV = "escala_cafe.csv";

    static final String[] INSTRUCOES = {
        "1  Desligue a garrafa de café",
        "2  Retire o pó usado e o café que sobrou do dia anterior",
        "3  Lave o coador",
        "4  Complete com água filtrada até a marca MAX (6L) da garrafa",
        "5  Insira o pino e o coador novamente na garrafa",
        "6  Coloque 16 colheres cheias de pó de café no coador (para 6L)",
        "7  Feche e ligue a cafeteira",
        "8  Informe o horário que ficará pronto (~30 min)",
    };

    static final String MARCAS       = "3 Corações (Forte, Extra Forte, Tradicional)  •  Melita";
    static final String INGREDIENTES = "Pó de café · 16 medidas · Água 6L (MAX) · ~30 min";

    // ── Helpers ────────────────────────────────────────────────────────────────
    static final String[] DIAS_PT  = {"Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"};
    static final String[] MESES_PT = {"jan","fev","mar","abr","mai","jun","jul","ago","set","out","nov","dez"};

    static final DateTimeFormatter FMT_FULL  = DateTimeFormatter.ofPattern("dd/MM/yyyy");
    static final DateTimeFormatter FMT_SHORT = DateTimeFormatter.ofPattern("dd/MM/yy");

    // ── Carregamento da escala ─────────────────────────────────────────────────
    /**
     * Lê escala_cafe.csv e retorna lista de Object[]{nome, data, pacotes}.
     * O CSV deve ter cabeçalho: nome,data,pacotes
     * Linhas em branco ou com nome vazio são ignoradas.
     */
    static List<Object[]> carregarEscala() throws IOException {
        List<Object[]> lista = new ArrayList<>();
        File arquivo = new File(ARQUIVO_CSV);
        if (!arquivo.exists()) {
            throw new IOException("Arquivo não encontrado: " + ARQUIVO_CSV +
                "\nExecute o cafe.bat para converter a planilha automaticamente.");
        }
        try (BufferedReader br = new BufferedReader(
                new InputStreamReader(new FileInputStream(arquivo), StandardCharsets.UTF_8))) {
            String cabecalho = br.readLine(); // pula cabeçalho
            if (cabecalho == null) return lista;
            String linha;
            while ((linha = br.readLine()) != null) {
                linha = linha.trim();
                if (linha.isEmpty()) continue;
                String[] partes = linha.split(",", -1);
                if (partes.length < 3) continue;
                String nome = partes[0].trim();
                String data = partes[1].trim();
                if (nome.isEmpty() || data.isEmpty()) continue;
                int pacotes = 0;
                try { pacotes = Integer.parseInt(partes[2].trim()); }
                catch (NumberFormatException ignored) {}
                lista.add(new Object[]{nome, data, pacotes});
            }
        }
        return lista;
    }

    // ── Formatação ─────────────────────────────────────────────────────────────
    static void limpar() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
    }

    static String repetir(String s, int n) {
        StringBuilder sb = new StringBuilder(n * s.length());
        for (int i = 0; i < n; i++) sb.append(s);
        return sb.toString();
    }

    static String linha(char ch, int n) {
        return repetir(String.valueOf(ch), n);
    }

    static String centralizar(String texto, int largura) {
        int pad = Math.max((largura - texto.length()) / 2, 0);
        return repetir(" ", pad) + texto;
    }

    static LocalDate parseData(String s) {
        for (DateTimeFormatter fmt : new DateTimeFormatter[]{FMT_FULL, FMT_SHORT}) {
            try { return LocalDate.parse(s, fmt); } catch (DateTimeParseException ignored) {}
        }
        return null;
    }

    static String formatarData(LocalDate dt) {
        int dow = dt.getDayOfWeek().getValue() - 1;
        return String.format("%02d/%s/%d  (%s)",
            dt.getDayOfMonth(),
            MESES_PT[dt.getMonthValue() - 1],
            dt.getYear(),
            DIAS_PT[dow]);
    }

    static String pacotesLabel(int n) {
        if (n == 0) return "0 pacotes doados :(";
        if (n == 1) return "1 pacote doado";
        return n + " pacotes doados";
    }

    // ── ASCII Art ──────────────────────────────────────────────────────────────
    static final String HEADER =
        "\n" +
        "                        (\n" +
        "                          )     (\n" +
        "                   ___...(-------)-....___\n" +
        "               .-\"\"       )    (          \"\"-.\n" +
        "         .-'``'|-._             )         _.-|\n" +
        "        /  .--.|   `\"\"---...........---\"\"`   |\n" +
        "       /  /    |                             |\n" +
        "       |  |    |                             |\n" +
        "        \\  \\   |           Bom dia!          |\n" +
        "         `\\ `\\ |             FSFX            |\n" +
        "           `\\ `|                             |\n" +
        "           _/ /\\                             /\n" +
        "          (__/  \\                           /\n" +
        "       _..---\"\"` \\                         /`\"\"---.._\n" +
        "    .-'           \\                       /          '-.\n" +
        "   :               `-.__             __.-'              :\n" +
        "   :                  ) \"\"---...---\"\" (                 :\n" +
        "    '._               `\"--...___...--\"`              _.'\n" +
        "      \\\"\"--..__ (                           ) __..--\"\"/\n" +
        "       '._     \"\"\"----.....______.....----\"\"\"     _.'\n" +
        "          `\"\"--..,,_____            _____,,..--\"\"`\n" +
        "                        `\"\"\"----\"\"\"`\n";

    static final String[] TITULO = {
        "╔══════════════════════════════════════════════════════════════════╗",
        "║      ☕  ESCALA DE CAFÉ  —  FUNDAÇÃO SÃO FRANCISCO XAVIER  ☕   ║",
        "║                    REV. 1  ·  03/26 - 06/26                      ║",
        "╚══════════════════════════════════════════════════════════════════╝",
    };

    static void imprimirAscii(String texto) {
        String[] linhas = texto.split("\n", -1);
        int maxLen = 0;
        for (String l : linhas) if (l.trim().length() > maxLen) maxLen = l.trim().length();
        int offset = Math.max((68 - maxLen) / 2, 0);
        for (String l : linhas) System.out.println(repetir(" ", offset) + l);
    }

    static void imprimirCabecalho() {
        limpar();
        imprimirAscii(HEADER);
        for (String t : TITULO) System.out.println(centralizar(t, 68));
        System.out.println();
    }

    static void imprimirInstrucoes() {
        System.out.println(centralizar("┌─────────────────────────────────────┐", 68));
        System.out.println(centralizar("│   PROCEDIMENTO PARA FAZER O CAFÉ    │", 68));
        System.out.println(centralizar("└─────────────────────────────────────┘", 68));
        System.out.println();
        for (String passo : INSTRUCOES) System.out.println("    " + passo);
        System.out.println();
        System.out.println("  Marcas    : " + MARCAS);
        System.out.println("  Info      : " + INGREDIENTES);
        System.out.println();
        System.out.println(linha('─', 68));
    }

    // ── Busca ──────────────────────────────────────────────────────────────────
    static void buscarPorNome(List<Object[]> escala, String termo) {
        termo = termo.trim().toLowerCase();
        List<Object[]> resultados = new ArrayList<>();
        for (Object[] entrada : escala) {
            if (((String) entrada[0]).toLowerCase().contains(termo))
                resultados.add(entrada);
        }
        if (resultados.isEmpty()) {
            System.out.printf("%n  ✗  Nenhum resultado para \"%s\".%n%n", termo);
            return;
        }
        System.out.printf("%n  ✔  %d entrada(s) encontrada(s):%n%n", resultados.size());
        System.out.printf("  %-22s %-30s %s%n", "NOME", "DATA", "PACOTES");
        System.out.println("  " + linha('-', 60));
        for (Object[] e : resultados) {
            LocalDate dt   = parseData((String) e[1]);
            String dataFmt = dt != null ? formatarData(dt) : (String) e[1];
            System.out.printf("  %-22s %-30s %s%n", e[0], dataFmt, pacotesLabel((int) e[2]));
        }
        System.out.println();
    }

    static void buscarPorData(List<Object[]> escala, String entrada) {
        entrada = entrada.trim();
        LocalDate dtBusca = parseData(entrada);
        if (dtBusca == null) {
            try {
                dtBusca = LocalDate.parse(entrada + "/2026", FMT_FULL);
            } catch (DateTimeParseException e) {
                System.out.println("\n  ✗  Data inválida. Use o formato DD/MM ou DD/MM/AAAA.\n");
                return;
            }
        }

        int dow = dtBusca.getDayOfWeek().getValue();
        if (dow >= 6) {
            String nomeDia = dow == 6 ? "SÁBADO" : "DOMINGO";
            System.out.printf("%n  ☕  Hoje é %s... café só na casa de mamãe! 🏠%n%n", nomeDia);
            return;
        }

        String dataFmt = formatarData(dtBusca);
        List<Object[]> encontrados = new ArrayList<>();
        for (Object[] e : escala) {
            LocalDate dt = parseData((String) e[1]);
            if (dt != null && dt.equals(dtBusca))
                encontrados.add(e);
        }

        if (encontrados.isEmpty()) {
            System.out.printf("%n  ✗  Nenhum responsável cadastrado para %s.%n%n", dataFmt);
            return;
        }

        System.out.printf("%n  📅  %s%n%n", dataFmt);
        System.out.printf("  %-25s %s%n", "RESPONSÁVEL", "PACOTES");
        System.out.println("  " + linha('-', 45));
        for (Object[] e : encontrados)
            System.out.printf("  %-25s %s%n", e[0], pacotesLabel((int) e[2]));
        System.out.println();
    }

    // ── Menu ───────────────────────────────────────────────────────────────────
    static void menu(int totalRegistros) {
        imprimirCabecalho();
        imprimirInstrucoes();
        System.out.printf("  Escala carregada: %d responsável(is)  ←  %s%n%n",
            totalRegistros, ARQUIVO_CSV);
        System.out.println("  PESQUISA");
        System.out.println("  " + linha('-', 40));
        System.out.println("  [1]  Buscar por NOME");
        System.out.println("  [2]  Buscar por DATA  (DD/MM ou DD/MM/AAAA)");
        System.out.println("  [0]  Sair");
        System.out.println();
    }

    // ── Main ───────────────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        System.setOut(new PrintStream(System.out, true, "UTF-8"));
        Scanner sc = new Scanner(System.in, "UTF-8");

        List<Object[]> escala;
        try {
            escala = carregarEscala();
        } catch (IOException ex) {
            System.out.println("\n  ERRO: " + ex.getMessage() + "\n");
            System.out.print("  Pressione ENTER para sair...");
            sc.nextLine();
            return;
        }

        if (escala.isEmpty()) {
            System.out.println("\n  AVISO: Nenhum dado encontrado em " + ARQUIVO_CSV + ".\n");
            System.out.print("  Pressione ENTER para sair...");
            sc.nextLine();
            return;
        }

        while (true) {
            menu(escala.size());
            System.out.print("  Opção: ");
            String opcao = sc.nextLine().trim();
            System.out.println();

            switch (opcao) {
                case "1":
                    System.out.print("  Digite o nome (parcial): ");
                    buscarPorNome(escala, sc.nextLine());
                    break;
                case "2":
                    System.out.print("  Digite a data (ex: 19/06 ou 19/06/2026): ");
                    buscarPorData(escala, sc.nextLine());
                    break;
                case "0":
                    limpar();
                    System.out.println("\n  Até logo! ☕\n");
                    return;
                default:
                    System.out.println("  ✗  Opção inválida.\n");
            }

            System.out.print("  Pressione ENTER para voltar ao menu...");
            sc.nextLine();
        }
    }
}

File: Python_café/cafe.py
# -*- coding: utf-8 -*-
import os
import sys
from datetime import datetime
from pathlib import Path

# ── Configuração da planilha ───────────────────────────────────────────────────
PASTA      = Path(__file__).parent
PLANILHA   = PASTA / "escala_cafe.xlsx"

# ── Carregamento dinâmico dos dados ───────────────────────────────────────────
def carregar_escala():
    try:
        from openpyxl import load_workbook
    except ImportError:
        print("\n  ✗  Biblioteca 'openpyxl' não encontrada.")
        print("     Instale com:  pip install openpyxl\n")
        sys.exit(1)

    if not PLANILHA.exists():
        print(f"\n  ✗  Planilha não encontrada: {PLANILHA}")
        print("     Certifique-se de que 'escala_cafe.xlsx' está na mesma pasta do script.\n")
        sys.exit(1)

    wb = load_workbook(PLANILHA, data_only=True)
    ws = wb.active
    escala = []
    for row in ws.iter_rows(min_row=2, values_only=True):
        nome, data, pacotes, concluido = (list(row) + [None, None, None, None])[:4]
        if not nome or not data:
            continue
        # Normaliza data: aceita objeto date/datetime ou string
        if hasattr(data, "strftime"):
            data = data.strftime("%d/%m/%Y")
        else:
            data = str(data).strip()
        pacotes   = int(pacotes)   if pacotes   is not None else 0
        concluido = bool(concluido) if concluido is not None else False
        escala.append((str(nome).strip(), data, pacotes, concluido))
    return escala

# ── Instruções ────────────────────────────────────────────────────────────────
INSTRUCOES = [
    "1  Desligue a garrafa de café",
    "2  Retire o pó usado e o café que sobrou do dia anterior",
    "3  Lave o coador",
    "4  Complete com água filtrada até a marca MAX (6L) da garrafa",
    "5  Insira o pino e o coador novamente na garrafa",
    "6  Coloque 16 colheres cheias de pó de café no coador (para 6L)",
    "7  Feche e ligue a cafeteira",
    "8  Informe o horário que ficará pronto (~30 min)",
]

MARCAS       = "3 Corações (Forte, Extra Forte, Tradicional)  •  Melita"
INGREDIENTES = "Pó de café · 16 medidas · Água 6L (MAX) · ~30 min"

# ── Helpers ────────────────────────────────────────────────────────────────────
DIAS_PT  = ["Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"]
MESES_PT = ["jan","fev","mar","abr","mai","jun","jul","ago","set","out","nov","dez"]

def limpar():
    os.system("cls" if os.name == "nt" else "clear")

def linha(char="─", n=68):
    return char * n

def centralizar(texto, n=68):
    return texto.center(n)

def parse_data(s):
    for fmt in ("%d/%m/%Y", "%d/%m/%y"):
        try:
            return datetime.strptime(s, fmt)
        except ValueError:
            pass
    return None

def formatar_data(dt):
    return f"{dt.day:02d}/{MESES_PT[dt.month-1]}/{dt.year}  ({DIAS_PT[dt.weekday()]})"

def colheres_label(n):
    if n == 0:
        return "0 pacotes doados :("
    elif n == 1:
        return "1 pacote doado"
    else:
        return f"{n} pacotes doados"

# ── ASCII Art ──────────────────────────────────────────────────────────────────
HEADER = r'''
                        (
                          )     (
                   ___...(-------)-....___
               .-""       )    (          ""-.
         .-'``'|-._             )         _.-|
        /  .--.|   `""---...........---""`   |
       /  /    |                             |
       |  |    |                             |
        \  \   |           Bom dia!          |
         `\ `\ |             FSFX            |
           `\ `|                             |
           _/ /\                             /
          (__/  \                           /
       _..---""` \                         /`""---.._
    .-'           \                       /          '-.
   :               `-.__             __.-'              :
   :                  ) ""---...---"" (                 :
    '._               `"--...___...--"`              _.'
      \""--..__                              __..--""/
       '._     """----.....______.....----"""     _.'
          `""--..,,_____            _____,,..--""`
                        `"""----"""`
'''

TITULO = [
    "╔══════════════════════════════════════════════════════════════════╗",
    "║      ☕  ESCALA DE CAFÉ  —  FUNDAÇÃO SÃO FRANCISCO XAVIER  ☕   ║",
    "║                    REV. 1  ·  03/26 - 06/26                      ║",
    "╚══════════════════════════════════════════════════════════════════╝",
]

def imprimir_ascii(texto):
    linhas = texto.splitlines()
    largura_max = max((len(l) for l in linhas if l.strip()), default=0)
    offset = max((68 - largura_max) // 2, 0)
    for l in linhas:
        print(" " * offset + l)

def imprimir_cabecalho(n_registros):
    limpar()
    imprimir_ascii(HEADER)
    for t in TITULO:
        print(centralizar(t))
    print(centralizar(f"[ planilha: escala_cafe.xlsx  ·  {n_registros} registros ]"))
    print()

def imprimir_instrucoes():
    print(centralizar("┌─────────────────────────────────────┐"))
    print(centralizar("│   PROCEDIMENTO PARA FAZER O CAFÉ    │"))
    print(centralizar("└─────────────────────────────────────┘"))
    print()
    for passo in INSTRUCOES:
        print(f"    {passo}")
    print()
    print(f"  {'Marcas    :'} {MARCAS}")
    print(f"  {'Info      :'} {INGREDIENTES}")
    print()
    print(linha())

# ── Busca ──────────────────────────────────────────────────────────────────────
def buscar_por_nome(escala, termo):
    termo = termo.strip().lower()
    resultados = [(n, d, c) for n, d, c, _ in escala if termo in n.lower()]
    if not resultados:
        print(f"\n  ✗  Nenhum resultado para \"{termo}\".\n")
        return
    print(f"\n  ✔  {len(resultados)} entrada(s) encontrada(s):\n")
    print(f"  {'NOME':<22} {'DATA':<30} {'PACOTES'}")
    print(f"  {linha('-', 60)}")
    for nome, data_str, col in resultados:
        dt = parse_data(data_str)
        data_fmt = formatar_data(dt) if dt else data_str
        print(f"  {nome:<22} {data_fmt:<30} {colheres_label(col)}")
    print()

def buscar_por_data(escala, entrada):
    entrada = entrada.strip()
    dt_busca = parse_data(entrada)
    if not dt_busca:
        try:
            dt_busca = datetime.strptime(entrada + "/2026", "%d/%m/%Y")
        except ValueError:
            print("\n  ✗  Data inválida. Use o formato DD/MM ou DD/MM/AAAA.\n")
            return

    dia_semana = dt_busca.weekday()
    if dia_semana >= 5:
        nome_dia = "sábado" if dia_semana == 5 else "domingo"
        print(f"\n  ☕  Hoje é {nome_dia.upper()}... café só na casa de mamãe! 🏠\n")
        return

    data_fmt = formatar_data(dt_busca)
    encontrados = []
    for nome, data_str, col, _ in escala:
        dt = parse_data(data_str)
        if dt and dt.date() == dt_busca.date():
            encontrados.append((nome, col))

    if not encontrados:
        print(f"\n  ✗  Nenhum responsável cadastrado para {data_fmt}.\n")
        return

    print(f"\n  📅  {data_fmt}\n")
    print(f"  {'RESPONSÁVEL':<25} {'PACOTES'}")
    print(f"  {linha('-', 45)}")
    for nome, col in encontrados:
        print(f"  {nome:<25} {colheres_label(col)}")
    print()

# ── Loop principal ─────────────────────────────────────────────────────────────
def main():
    while True:
        escala = carregar_escala()          # relê a planilha a cada iteração
        imprimir_cabecalho(len(escala))
        imprimir_instrucoes()
        print()
        print("  PESQUISA")
        print(f"  {linha('-', 40)}")
        print("  [1]  Buscar por NOME")
        print("  [2]  Buscar por DATA  (DD/MM ou DD/MM/AAAA)")
        print("  [0]  Sair")
        print()

        opcao = input("  Opção: ").strip()
        print()

        if opcao == "1":
            termo = input("  Digite o nome (parcial): ").strip()
            buscar_por_nome(escala, termo)

        elif opcao == "2":
            entrada = input("  Digite a data (ex: 19/06 ou 19/06/2026): ").strip()
            buscar_por_data(escala, entrada)

        elif opcao == "0":
            limpar()
            print("\n  Até logo! ☕\n")
            sys.exit(0)

        else:
            print("  ✗  Opção inválida.\n")

        input("  Pressione ENTER para voltar ao menu...")

if __name__ == "__main__":
    main()